src/Microsoft.VisualStudio.LanguageServices.Razor/TagHelperDescriptorJsonConverter.cs

```diff
@@ -0,0 +1,234 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Razor.Evolution;
+using Microsoft.AspNetCore.Razor.Evolution.Legacy;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    internal class TagHelperDescriptorJsonConverter : JsonConverter
+    {
+        public static readonly TagHelperDescriptorJsonConverter Instance = new TagHelperDescriptorJsonConverter();
+        private const string RazorDiagnosticMessageKey = "Message";
+        private const string RazorDiagnosticTypeNameKey = "TypeName";
+
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType == typeof(TagHelperDescriptor) ||
+                objectType == typeof(DefaultRazorDiagnostic) ||
+                objectType == typeof(LegacyRazorDiagnostic);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType != JsonToken.StartObject)
+            {
+                return null;
+            }
+
+            var descriptor = JObject.Load(reader);
+            var descriptorKind = descriptor[nameof(TagHelperDescriptor.Kind)].Value<string>();
+            if (descriptorKind != ITagHelperDescriptorBuilder.DescriptorKind)
+            {
+                throw new NotSupportedException();
+            }
+
+            var typeName = descriptor[nameof(TagHelperDescriptor.Name)].Value<string>();
+            var assemblyName = descriptor[nameof(TagHelperDescriptor.AssemblyName)].Value<string>();
+            var tagMatchingRules = descriptor[nameof(TagHelperDescriptor.TagMatchingRules)].Value<JArray>();
+            var boundAttributes = descriptor[nameof(TagHelperDescriptor.BoundAttributes)].Value<JArray>();
+            var childTags = descriptor[nameof(TagHelperDescriptor.AllowedChildTags)].Value<JArray>();
+            var documentation = descriptor[nameof(TagHelperDescriptor.Documentation)].Value<string>();
+            var tagOutputHint = descriptor[nameof(TagHelperDescriptor.TagOutputHint)].Value<string>();
+            var diagnostics = descriptor[nameof(TagHelperDescriptor.Diagnostics)].Value<JArray>();
+            var metadata = descriptor[nameof(TagHelperDescriptor.Metadata)].Value<JObject>();
+
+            var builder = ITagHelperDescriptorBuilder.Create(typeName, assemblyName);
+
+            builder
+                .Documentation(documentation)
+                .TagOutputHint(tagOutputHint);
+
+            foreach (var tagMatchingRule in tagMatchingRules)
+            {
+                builder.TagMatchingRule(b => ReadTagMatchingRule(b, tagMatchingRule.Value<JObject>(), serializer));
```

Comment:
Break these things up so they're easy to debug if we need to in the future. Same with the other loops below